issue#11 | Remove items like a PRO

saraheisa opened the pull request:
# Description 

This PR enables the user to remove items from their list & Create a FlashMessage component to show success/error messages to the user

### Remove Items from the list

- Added a delete icon using [font awesome cdn](https://fontawesome.com) to each item in the list
- Added a confirmation prompt before deletion
- Displayed a success message after successful deletion or error message after failed deletion

### Create FlashMessage Component

- Reused the flash message in `AddItem` component
- Created `Bus` an `EventEmitter` object that's used to set flash messages from different components using `Bus.emit()`

## Related Issue

#11 

## Acceptance Criteria

- [X] User is able to delete an item from the shopping list
- [X] Before deleting, prompt the user to confirm that they really want to delete the item to prevent accidental deletions
- [X] Deletion should cause the associated record(s) in the database to be deleted

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

![deleteitem](https://user-images.githubusercontent.com/14318900/98735729-8ce38480-23ac-11eb-9e2b-293de60c016d.gif)

## Testing Steps / QA Criteria

- From your terminal, pull down this branch with git pull origin `se-dg-issue-11`
- Check that branch out with `git checkout se-dg-issue-11`
- Run `yarn start` to run the app
- Create a new list
- Go to `add` and add some items!!
- Go back to the `list` and you should see the items you just added with a bin icon on their right side
- Choose your least favorite item and click on the bin next to it. Now your browser will yell at you with a confirmation dialog whether you want to delete the item or not choose `Ok!`
- Then you should see a confirmation message that the item has been deleted successfully and you should notice that the item has disappeared from the list

deboragaleano committed:
feat: add remove functionality to list items

saraheisa committed:
Feat: Create Flash Message components

this component can be used to show success/error messages to the user
it uses Bus.js to set messages

saraheisa committed:
Refactor: Use FlashComponent to show messages to user

saraheisa committed:
Feat: Use FlashMessages to show messages to the user

Show success message when an item is deleted successfully
And an error message when something goes wrong

deboragaleano committed:
style: add font-awesome

saraheisa committed:
Style: Add styles to ListItem

adjust colors and spacing

saraheisa committed:
Style: add pointer to delete icon

saraheisa committed:
Refactor: Use button instead of span for A11Y

add aria-label
adjust styles for the button

saraheisa committed:
Feat: add aria-label to checkbox

saraheisa committed:
Fix: Remove unused property and set timeout back to 2ms

I hope no one notice this!!

deboragaleano committed:
fix MaxListenersExceededWarning by removing listeners in useEffect

saraheisa committed:
Refactor: Change let to const for timeoutId and add comments

hajarNasr commented on src/components/ListItem.js:
```diff
@@ -8,6 +11,30 @@ import { userShoppingList } from '../lib/shoppingListsCollection';
+          [itemId]: firebase.firestore.FieldValue.delete(),
+        })
+        .then((res) => {
+          Bus.emit('flash', {
+            content: `${listItem.name} deleted successfully`,
+            type: 'success',
+          });
+        })
+        .catch((res) => {
+          Bus.emit('flash', {
+            content: `Something went wrong, Sorry!!!`,
+            type: 'error',
+          });
```
I didn't know about `EventEmitter` and `events` so I had to google them. However, I loved that you used them to show different types of messages to the user. ❤️

hajarNasr commented on src/components/flashMessage.js:
```diff
@@ -0,0 +1,34 @@
+import React, { useState, useEffect } from 'react';
+import Bus from '../lib/bus';
+
+const FlashMessage = () => {
```
Love this component!

hajarNasr commented on src/components/flashMessage.js:
```diff
@@ -0,0 +1,34 @@
+      displayMessage(content, type);
+    });
+
+    const timeoutId = setTimeout(() => setMessage(intialMessage), 2000);
```
I know that this line was in a different issue, but now I've come to realize that maybe we need to check `if (message.content)` before setting the msg? I think `useEffect` is going to run and the message is going to be reset every two seconds if we don't check.

saraheisa replied:
Yup! you're so right 
we didn't even notice this!
Kudos to you 🙌

hajarNasr commented on src/components/flashMessage.js:
```diff
@@ -0,0 +1,34 @@
+  const [message, setMessage] = useState(intialMessage);
+
+  useEffect(() => {
+    Bus.addListener('flash', ({ content, type }) => {
+      displayMessage(content, type);
+    });
```
I think we can add this listener in another `useEffect` that has an empty array as a dependency so that we only `addListener` and `removeAllListeners` once. :)

saraheisa replied:
yeah makes sense
Thank youu 👍

hajarNasr commented on src/components/ListItem.js:
```diff
@@ -42,8 +69,16 @@ const ListItem = ({ listItem, itemId }) => {
         onChange={() => checkItem()}
         disabled={isChecked}
         checked={isChecked}
+        aria-label="Mark this item as purchased"
```
Love that you added aria labels even though that wasn't in the AC of your issue. 👍

yahayakenny replied:
Great work guys! Everything works well here. Nicely done 👍😊

hajarNasr reviewed (commented):
Everything worked perfectly on my end. I was able to delete items and the deletion was reflected on the database. So great job. 💯

saraheisa committed:
Refactor: add listener to bus only once

create a useEffect with an empty array dependancy to make it called once
add listener to bus as the useEffect
and remove listeners from the bus in the cleanup function

saraheisa committed:
Fix: Prevent useEffect (with setTimeout) from being called constantly

Checking if message content is empty before calling setMessage
because everytime setMessage is called the message state will change
and as a result useEffect will be called which will call setMessage... and so on
like an infinite misery loop

saraheisa committed:
Fix: Use aria-live instead of role

Turns out that role doesn't work as expected on screen readers
the screen reader only said "alert" even with aria-label
but aria-live works as expected with aria-label
Giving the value "rude" to aria-live is not rude!! it's used to interrupt the screen
reader of whatever it's doing to read the message
so it's used for the user's own sake

llunaplanet commented on src/components/AddItem.js:
```diff
@@ -38,10 +27,10 @@ const AddItem = () => {
       .then((data) => {
         if (data.data()) {
           if (existingName(data.data(), newItem.name)) {
-            displayMessage(
```
What do you think about instead of removing **displayMessage**, move the **Bus.emit** logic inside the displayMessage function. This can have 2 benefits, 

1. If at some point you want to change the implementation, you can keep the interface and only change the **displayMessage** function ( touching only one file / function ). 
2. About semantics, **displayMessage** vs **Bus.emit('flash')**, the first seems clearer

asianvader replied:
I really like the idea of using an EventEmitter to display your flash messages. But it would be even better - as David suggested - that you move the Bus.emit inside the displayMessage function.

saraheisa replied:
Yeah agree with you that will make it more readable

llunaplanet commented on src/components/flashMessage.js:
```diff
@@ -0,0 +1,41 @@
+  useEffect(() => {
+    const timeoutId = setTimeout(() => {
+      if (message.content) setMessage(intialMessage);
+    }, 2000);
```
What do you think about giving more semantics to the timeout time [2000]? We usually call this **Magic numbers**,  https://en.wikipedia.org/wiki/Magic_number_(programming)

saraheisa replied:
Thanks a lot for the valuable suggestions :))

I wonder why it's called magic number it's just a constant?! This industry is filled with buzzwords 😅

llunaplanet reviewed (approved):
Great work! All the AC has been meet, great to see inter-component communication, and it looks you had fun with setTimeout 😁. Well done I left some non blocking comments

asianvader commented on src/components/flashMessage.js:
```diff
@@ -0,0 +1,41 @@
+
+    // remove all the listeners from EventEmitter(Bus) so we don't have a memory leak!!
+    return () => {
+      Bus.removeAllListeners('flash');
```
Nice! I like a tidy app! :) Good you're thinking about performance.

asianvader commented on src/components/ListItem.js:
```diff
@@ -42,8 +69,16 @@ const ListItem = ({ listItem, itemId }) => {
+      <button
+        onClick={removeItem}
+        className="delete-icon"
+        aria-label="Delete this item from your shopping list"
```
Good that you added the aria-label attribute for accessibility.

asianvader reviewed (approved):
Nice work here. Nothing blocking, just comments and suggestions. Works very well. Noticed you couldn't resist making your app pretty with the use of font awesome icons! ^.^

Lukabaramishvili commented on public/index.html:
```diff
@@ -24,6 +24,7 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
+    <script src="https://kit.fontawesome.com/b4b59ecaeb.js"></script>
```
I was wondering if this imports only one, delete icon?

saraheisa replied:
It's a new way to use font awesome [font awesome kits](https://blog.fontawesome.com/introducing-font-awesome-kits)
Not sure if this only import the delete icon, but I have checked the network tab and there're some requests for CSS files!!
Anyway they say this way is faster so we're just going to trust them

Lukabaramishvili replied:
Yep, I've never used font awesome with kit. 
ah, this link is helpful. It says, `Automatic Subsetting! Speed up your web sites by loading only the icons your site is using.` cool cool! 👌🏻

saraheisa replied:
Oh I haven't notice this line 
Nice!

Lukabaramishvili reviewed (approved):
Everything works as expected! Great work! 👏 Just left one question 👍

deboragaleano committed:
refactor: give semantics to timeout: change seconds to magic number

deboragaleano committed:
fix: add defer to fontawesome

deboragaleano committed:
refactor: add displayMessage as a helper function
